File: src/main/java/org/example/expert/domain/todo/controller/TodoController.java
package org.example.expert.domain.todo.controller;

import java.time.LocalDate;
import java.time.LocalDateTime;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.example.expert.domain.common.dto.AuthUser;
import org.example.expert.domain.todo.dto.request.TodoSaveRequest;
import org.example.expert.domain.todo.dto.response.TodoResponse;
import org.example.expert.domain.todo.dto.response.TodoSaveResponse;
import org.example.expert.domain.todo.dto.response.TodoSummaryResponse;
import org.example.expert.domain.todo.service.TodoService;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PageableDefault;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
public class TodoController {

    private final TodoService todoService;

    @PostMapping("/todos")
    public ResponseEntity<TodoSaveResponse> saveTodo(
            @AuthenticationPrincipal AuthUser authUser,
            @Valid @RequestBody TodoSaveRequest todoSaveRequest
    ) {
        return ResponseEntity.ok(todoService.saveTodo(authUser, todoSaveRequest));
    }

    @GetMapping("/todos")
    public ResponseEntity<Page<TodoResponse>> getTodos(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int size
    ) {
        return ResponseEntity.ok(todoService.getTodos(page, size));
    }

    @GetMapping("/todos/search")
    public ResponseEntity<Page<TodoResponse>> searchTodos(
        @RequestParam(defaultValue = "1") int page,
        @RequestParam(defaultValue = "10") int size,
        @RequestParam(required = false) String weather,
        @RequestParam(required = false) String start,
        @RequestParam(required = false) String end
    ) {
        LocalDateTime startDateTime = (start != null) ? LocalDateTime.parse(start) : null;
        LocalDateTime endDateTime = (end != null) ? LocalDateTime.parse(end) : null;
        return ResponseEntity.ok(todoService.searchTodos(page, size, weather, startDateTime, endDateTime));
    }

    @GetMapping("/todos/{todoId}")
    public ResponseEntity<TodoResponse> getTodo(@PathVariable long todoId) {
        return ResponseEntity.ok(todoService.getTodo(todoId));
    }

    @GetMapping("/todos/searchFilters")
    public ResponseEntity<Page<TodoSummaryResponse>> searchTodosWithFilters(
        @RequestParam(required = false) String keyword,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)LocalDate startDate,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)LocalDate endDate,
        @RequestParam(required = false) String nickname,
        @PageableDefault(size = 10, sort = "createdAt", direction = Sort.Direction.DESC) Pageable pageable
    ) {
        Page<TodoSummaryResponse> result = todoService.searchTodosWithFilters(keyword, nickname, startDate, endDate, pageable);
        return ResponseEntity.ok(result);
    }
}

File: src/main/kotlin/com/example/domain/todo/controller/TodoController.kt
package com.example.domain.todo.controller

import jakarta.validation.Valid
import com.example.domain.common.dto.AuthUser
import com.example.domain.todo.dto.request.TodoSaveRequest
import com.example.domain.todo.dto.response.TodoResponse
import com.example.domain.todo.dto.response.TodoSaveResponse
import com.example.domain.todo.dto.response.TodoSummaryResponse
import com.example.domain.todo.service.TodoService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PageableDefault
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import java.time.LocalDate
import java.time.LocalDateTime

@RestController
@RequestMapping("/todos")
class TodoController(
    private val todoService: TodoService

) {
    @PostMapping
    fun saveTodo(
        @AuthenticationPrincipal authUser: AuthUser,
        @RequestBody @Valid request: TodoSaveRequest
    ): ResponseEntity<TodoSaveResponse> {
        return ResponseEntity.ok(todoService.saveTodo(authUser, request))
    }

    @GetMapping
    fun getTodos(
        @RequestParam(defaultValue = "1") page: Int,
        @RequestParam(defaultValue = "10") size: Int
    ): ResponseEntity<Page<TodoResponse>> {
        return ResponseEntity.ok(todoService.getTodos(page, size))
    }

    @GetMapping("/search")
    fun searchTodos(
        @RequestParam(defaultValue = "1") page: Int,
        @RequestParam(defaultValue = "10") size: Int,
        @RequestParam(required = false) weather: String?,
        @RequestParam(required = false) start: String?,
        @RequestParam(required = false) end: String?
    ): ResponseEntity<Page<TodoResponse>> {
        val startDateTime = start?.let { LocalDateTime.parse(it) }
        val endDateTime = end?.let { LocalDateTime.parse(it) }
        return ResponseEntity.ok(todoService.searchTodos(page, size, weather, startDateTime, endDateTime))
    }

    @GetMapping("/{todoId}")
    fun getTodo(@PathVariable todoId: Long): ResponseEntity<TodoResponse> {
        return ResponseEntity.ok(todoService.getTodo(todoId))
    }

    @GetMapping("/searchFilters")
    fun searchTodosWithFilters(
        @RequestParam(required = false) keyword: String?,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) startDate: LocalDate?,
        @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) endDate: LocalDate?,
        @RequestParam(required = false) nickname: String?,
        @PageableDefault(size = 10, sort = ["createdAt"], direction = Sort.Direction.DESC) pageable: Pageable
    ): ResponseEntity<Page<TodoSummaryResponse>> {
        val result = todoService.searchTodosWithFilters(keyword, nickname, startDate, endDate, pageable)
        return ResponseEntity.ok(result)
    }
}
